pkg/infoschema/builder.go

```diff
@@ -112,60 +112,133 @@ func (b *Builder) applyCreateTables(m meta.Reader, diff *model.SchemaDiff) ([]in
-		schemaDiff := &model.SchemaDiff{
-			Version:  diff.Version,
-			Type:     model.ActionCreateSchema,
-			SchemaID: schemaID,
+
+		// Schema doesn't exist in kv, drop it from infoschema.
+		if dbInfo == nil {
+			schemaDiff := &model.SchemaDiff{
+				Version:     diff.Version,
+				Type:        model.ActionDropSchema,
+				SchemaID:    schemaID,
+				OldSchemaID: oldSchemaID,
+				OldTableID:  oldTableID,
+			}
+			return applyDropSchema(b, schemaDiff), nil
 		}
-		if err := applyCreateSchema(b, m, schemaDiff); err != nil {
-			return nil, errors.Trace(err)
+
+		// Schema exists in kv but not in infoschema, create it to infoschema
+		if _, ok := b.SchemaByID(schemaID); !ok {
+			schemaDiff := &model.SchemaDiff{
+				Version:     diff.Version,
+				Type:        model.ActionCreateSchema,
+				SchemaID:    schemaID,
+				OldSchemaID: oldSchemaID,
+				OldTableID:  oldTableID,
+			}
+			if err := applyCreateSchema(b, m, schemaDiff); err != nil {
+				return nil, errors.Trace(err)
+			}
+		} else {
+			currentSchema, _ := b.SchemaByID(schemaID)
+
+			needsCharsetUpdate := currentSchema.Charset != dbInfo.Charset || currentSchema.Collate != dbInfo.Collate
+			needsPlacementUpdate := !equalPlacementPolicy(currentSchema.PlacementPolicyRef, dbInfo.PlacementPolicyRef)
+
+			if needsCharsetUpdate {
+				charsetDiff := &model.SchemaDiff{
+					Version:     diff.Version,
+					Type:        model.ActionModifySchemaCharsetAndCollate,
+					SchemaID:    schemaID,
+					OldSchemaID: oldSchemaID,
+					OldTableID:  oldTableID,
+				}
+				if err := applyModifySchemaCharsetAndCollate(b, m, charsetDiff); err != nil {
+					return nil, errors.Trace(err)
+				}
+			}
+
+			if needsPlacementUpdate {
+				placementDiff := &model.SchemaDiff{
+					Version:     diff.Version,
+					Type:        model.ActionModifySchemaDefaultPlacement,
+					SchemaID:    schemaID,
+					OldSchemaID: oldSchemaID,
+					OldTableID:  oldTableID,
+				}
+				if err := applyModifySchemaDefaultPlacement(b, m, placementDiff); err != nil {
+					return nil, errors.Trace(err)
+				}
+			}
 		}
+		return nil, nil
 	}
 
-	// If tableID is not specified, skip refresh meta.
-	if tableID <= 0 {
+	// Table operation - check if database exists in infoschema first
+	// if not just return, since infoschema is consistent, if schema is gone then all tables must have gone.
+	if _, ok := b.SchemaByID(schemaID); !ok {
 		return nil, nil
 	}
-	// Check if table exists in kv.
+
+	// Check if table exists in kv
 	tableInfo, err := m.GetTable(schemaID, tableID)
-	if err != nil {
+	if err != nil && !meta.ErrDBNotExists.Equal(err) {
 		return nil, errors.Trace(err)
 	}
-	// Table not exists in kv, drop it from infoschema.
+
+	// Table not exists in kv, drop it from infoschema
 	if tableInfo == nil {
 		schemaDiff := &model.SchemaDiff{
-			Version:  diff.Version,
-			Type:     model.ActionDropTable,
-			SchemaID: schemaID,
-			TableID:  tableID,
+			Version:     diff.Version,
+			Type:        model.ActionDropTable,
```

Comment:
will the refresh meta handle the partition rather than table? if so, do we need modify the schemadiff.Type base on table/partion?